Fix All of RTL (#3342)

tywrenn opened the pull request:
Fixes #3342

Recreated this pull since I messed up when merging master

#### Short description of what this resolves:

Fixes Patient Portal RTL functionality

**Note/Edit: After further analysis, it has come to my attention that most if not all of RTL needs improving. Expect major changes to RTL with this pull.**

#### Changelog so far

- Converts the rest of the portal styles into SCSS files
- Adds the new portal RTL mixin to override some things for RTL
- Fully redesigns the Patient Portal navbar to use what BS4 really wanted us to use
- (Unrelated) Fixes the Patient Portal login so that it actually shows what's causing an error when attempting to login
- Other fixes


below added by bradymiller:
Up For Grabs for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Delta_-_Up_For_Grabs_Demo

tywrenn commented:
@sjpadgett The problem with not being able to login to the portal is that now even after reverting to master I can't login to it. Also checked the error log -- no errors. Seems like SQL is really hardcore being messed up here.

bradymiller commented:
hi @tywrenn , At least it is secure :)
Are you setting the email address (both normal one and secure one) for the patient? This is something I sometimes forget to set.

tywrenn commented:
@bradymiller yep I actually have the switch on for emails as usernames

sjpadgett commented:
> yep I actually have the switch on for emails as usernames

email for username is default for new credentials. The switch is for forcing to have email entered as part of login. You need to match patients email in demographics. Also, username is not case sensitive.

tywrenn commented:
This commit is a big one. I figured out RTL didn't really work too well after playing around with it. So here's a huge change to RTL:
- Changed out the Bootstrap RTL vendor (the other one didn't work well)
- Added a gulp string replacer that will add Bootstrap inside of SASS and swap between the RTL version and the regular one (there was no other way to do this)
- We weren't using directional.scss to its full potential. This fixes that by adding variables in place of left and right
- The calendar wasn't using RTL at all. This fixes that completely.
- This moves a number of regular CSS files to SASS (we were gonna do that anyways)
- Changes the entirety of how nav tabs CSS is compiled. Now it's SASS and works better for RTL.
- Reverts changes on get_patient_info.php and index.php in patient portal (I can't login still, this needs to be fixed)
- Other fixes

Let me know if something broke in this commit because i'm sure that something might've in RTL.

tywrenn commented:
This pull is still being worked on yet the changes that I'm making will add some new styling guidelines to our styling guide when this is all done:

When writing to any SASS file, you must keep in mind that OpenEMR has Right to Left for other languages. So if you are using any variable/attribute that has the word `left` or `right` in it, you must use the following guidelines:
- Anything formatted as e.g. `float: right;` must be changed to `float: $right;` to allow RTL support
- If you are specifying margin left/right or padding left/right and both sides aren't the same values or you only have one e.g. left attribute set, you must format it like this for example: `margin-#{$left}: 5px;`. This allows for right to left support.

tywrenn commented:
This pull is now on draft as we need to solve the portal login issue before I even continue working on this

sjpadgett commented:
@tywrenn 
Please note that you can't include opener in portal ie Header or otherwise. Opener is placed in Top by ref. Portal has a dummy top.opener in Home. Plus dialog has several hooks to ensure it can close.

tywrenn commented:
@sjpadgett Where is this referenced at?

sjpadgett commented:
> If ur talking about site_id

I'm talking about include_opener.js or Header(['opener']) etc.
One place i've noticed is portal/patient/templates/_Header.tpl.php

Also site_id does get used or multi site won't work!

tywrenn commented:
@sjpadgett Well good thing I figured out site_id was broken 😂 it was on some old getter that did nothing lol. Correct me if I'm wrong but I don't believe I changed that file in this pull as far as I can see.

tywrenn commented:
This still is a draft right now as I still need to fix the collapse part of the portal navbar. The last commit changed the following:
- Reverted sidebar. It didn't really work that well.
- Removed the sidebar in the onsite document view in favor of a second navbar.
- Reset home.php, refixed the Font Awesome icons, and converted the design to a flex view instead.
- In the process of changing the sidebar collapse to something more easier with less JavaScript/jQuery
- Readds multi-site support variable -- it was broken before for some reason
- Other fixes

sjpadgett commented:
I'm not sure if I really care too much concerning sidebars. I can take or leave um, I guess.

tywrenn commented:
@sjpadgett This is why I removed the onsite document list view's sidebar -- there was too much JavaScript going on in that and was just not reliable. With home's sidebar, since we have so many buttons, i'm trying to do everything to keep it and it looks like I will have success (plus RTL looks SOOOOOOOOOOO much better right now)

sjpadgett commented:
Okay Ty. Do what ya think is best but, you know that Patient Documents is my little pet and very protective of it.:)
Also be aware that Patient Documents also has it's counterpart in Patient Summary Documents so solution should also have a clean UI there as well...

sjpadgett commented:
Very nice @tywrenn 
Losing sidebar in patient documents is good move. RTL looks great.
I only noted a couple html syntax errors in home page but i'm sure you're prob not done but, you sure are close.
thanks

tywrenn commented:
Rebasing stinks yet I just did that

tywrenn commented:
Changelog:
- Fixes critical syntax errors
- Fixes register.php issues and adds portal style
- Overhauls appointment system, now runs better than it did before
- Converts some areas to BS4 compliance
- Adds portal style to messages area
- Fixes broken headers
- Other fixes

This pull is now out of draft and is ready for testing and review.

sjpadgett commented:
@tywrenn 
Noted RTL works great in portal however, broken in core.

tywrenn commented:
@sjpadgett Shouldn't be broken -- tested this a crap ton. Where specifically is it broken?

sjpadgett commented:
hmm, let me rebuild again. must be something with my install.
![image](https://user-images.githubusercontent.com/1859985/81459291-2cf5f200-916d-11ea-869e-446bf4944ca1.png)

sjpadgett commented:
@bradymiller Can you get this reset on a up for grab. I can't get RTL to work in main on windows. Only portal!
@tywrenn Any chance you missed a file for PR?

bradymiller commented:
Up For Grabs for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Delta_-_Up_For_Grabs_Demo

tywrenn commented:
Just noticed the navbar in RTL mode when collapsed is funky. Will fix this when I move on to fix the entire navbar.

bradymiller commented on portal/get_patient_info.php:
```diff
@@ -21,11 +21,6 @@
-//
-
-// landing page definition -- where to go if something goes wrong
-$landingpage = "index.php?site=" . urlencode($_SESSION['site_id']);
```
This change seems really odd since then the calls that use this $landingpage variable below will not be defined???

tywrenn replied:
I don't see anywhere that calls this method in a get request so I changed it to something more practical

bradymiller replied:
But then it will not be defined where it is used in this script in lines 28, 35, and 41. (Note the code here was not changed (you did the GET to SESSSION in another script which made sense there), just moved here)

tywrenn replied:
Ohhhhh now I see what ur talking about 😂 Yeah I'll go ahead and change that

tywrenn replied:
Just realized that this variable is actually a dummy and doesn't actually go to any part of the code!!!

bradymiller commented:
hi @tywrenn , Very nice stuff! Only noted one issue on the code review that posted above. Otherwise looks really nice and the core OpenEMR RTL testing is looking good on the Up For Grabs demo. After address above issue and ensure things are working on @sjpadgett 's end of things, then I think this is ready for the codebase.

sjpadgett commented on portal/home.php:
```diff
@@ -543,72 +536,78 @@ function startAssessment(param, assessmentOID) {
                     </div>
             </section>
-        </aside><!-- /.right-side -->
+        </main><!-- /.right-side -->
```
Mismatched end tag `<aside>`

sjpadgett commented on portal/patient/templates/_Header.tpl.php:
```diff
@@ -14,7 +14,7 @@
 <!DOCTYPE html>
-<html lang="en">
+<html>
     <head>
         <meta charset="utf-8" />
 
```
`<?php Header::setupHeader(['no_main-theme', 'patientportal-style', 'opener', 'fontawesome', 'datetime-picker']); ?>`
Needs to be fixed. Portal hasn't anywhere to put 'opener'. remove.

Also, remove Lab script loads:
.script("<?php echo $GLOBALS['assets_static_relative']; ?>/jquery/dist/jquery.min.js").wait()
                .script("<?php echo $GLOBALS['assets_static_relative']; ?>/bootstrap/dist/js/bootstrap.bundle.min.js")
                .script("<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-datetimepicker/build/jquery.datetimepicker.full.min.js")

Otherwise we duplicate loading dependencies.

tywrenn replied:
Glad we're fixing this part 😂 the headers in all these templates are pretty terrible

sjpadgett replied:
be careful with headers as there is a reason some are as they are.
There can be timing issues and loading order that is important.

Just leave these with me and i'll do follow up PR. There are other issues that I need to address as well.

sjpadgett commented:
@tywrenn 
Let me know when you're ready and i'll bring in. This PR is becoming hard to manage plus I have a security PR to get in that may conflict here.

Follow up needs
- dialogs broken due to past overflow removal.
- unclear if we mean for home sidebar to not be sticky.
- signature semi broken in dashboard document signing. okay elsewhere
- portal edit/create appointment needs to be restyled for responsive. it may be better to lose the dialog and create a collapsible div.

I'm extremely pleased with portal Tyler. Nice job so thank you.

tywrenn commented:
@sjpadgett ready to merge

sjpadgett commented:
Okay, i'm releasing this into the wilds! Thanks again for the hard work...

bradymiller commented on portal/get_patient_info.php:
```diff
@@ -22,6 +22,22 @@
+if (isset($_GET['site'])) {
+    $landingpage = "index.php?site=" . urlencode($_GET['site']);
+}
+
```
This makes me nervous. Was this to fix a bug? I'll look into this a bit closer.

bradymiller replied:
Yep, will prob revert the changes in this file since need to block these things rather than give access (via the $ignoreAuth = 1; and globals.php call)
